modules/core/src/main/scala/tethys/package.scala

```diff
@@ -44,7 +44,8 @@ package object tethys {
 
   implicit class ReaderReaderOps(val reader: Reader) extends AnyVal {
     def readJson[A](implicit jsonReader: JsonReader[A], producer: TokenIteratorProducer): Either[ReaderError, A] = {
-      producer.fromReader(reader).readJson[A]
+      implicit val root: FieldName = FieldName()
+      ReaderError.catchNonFatal(producer.fromReader(reader)).right.flatMap(_.readJson[A])
```

Comment:
This look much better, but still slightly weird.
I Explore your suggestion to change fromReader signature but it's seems to be a little bit complicated :D